### ISSUE TITLE
 [css-nav-1] Consider candidates on the navigation direction.

### ISSUE THREAD

**Issue by jeonghee27** [w3c/csswg-drafts#4483]
I want to resolve some unreachable cases caused by current spec about determining whether the candidate is on the navigation direction or not.

In the below example, Let's assume that user press right-button from A/C/E.

![image](https://user-images.githubusercontent.com/34676178/68172679-ccb6a580-ffbb-11e9-9795-7ff78e9d6e07.png)

B and D is reachable from A /C. 
According to current spec and chromium, F is not right side of E 

*Spec :
https://drafts.csswg.org/css-nav-1/ 
-> "the item partially overlaps with searchOrigin and its two edges which are orthogonal to dir should be on the navigation direction of the respective ones of searchOrigin."

*Chromium : https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/page/spatial_navigation.cc?q=spatial_navigation.cc&sq=package:chromium&dr&l=117
static inline bool RightOf(const PhysicalRect& a, const PhysicalRect& b) {
  return a.X() >= b.Right() || (a.X() >= b.X()  && a.Right() > b.Right() &&
                                a.Y() < b.Bottom() && a.Bottom() > b.Y());
}

I hope to remove "a.Right() > b.Right()" part.
Is there a problem if User can go from E to F using right key?
I would like to know your opinions. (@jihyerish @frivoal @hugoholgersson @Bokand @anawhj )

**Comment by anawhj** [comment]
I simplify the problem and propose the candidate solution via the following example.
https://software.hixie.ch/utilities/js/live-dom-viewer/?saved=7345
<img src="https://raw.githubusercontent.com/lgewst/images/master/partially_overlapped_issue.png">

The example above shows the current(As Is) behavior of SpatNav (polyfill and Chromium) as follows:

- As Is: A-B-E (pressing the right arrow key from A to E)
- To Be: A-B-C-D-E (pressing the right arrow key from A to E)

In the figure above, B is considered to be on the right of A, if **the left and right edges** of B are on the right of the respective ones of A as a rule specified in css-nav spec. For partially overlapping elements, D can't be a candidate if users press a right arrow key from B now. The reason is that the right edge of D isn't on the right of the right edge of B.

For securing D's reachability in a right direction from B, Jeonghee would suggest that we could modify the rule to ways only considering **the left edge** on partially overlapping elements when pressing a right arrow key.

It's definitely a heuristic rule that doesn't means a clear solution satisfying all people, but it definitely solves several abnormal behavior(e.g. unreachable element by arrow keys) in several general web sites such as espn.com and naver.com.

**Comment by jihyerish** [comment]
I think considering 'C' in the above example as a candidate, in general, is risky. 
From 'B', and when pressing the right arrow key, 'D' needs to be included in the set of candidates because it'll be unreachable in any direction.
But for 'C', it can be reachable anyway when pressing the up arrow key from 'B'.
It's true that it will solve the issue of https://naver.com, but I'm not sure it will be okay for the other pages. 

Therefore I've described how to select the partially overlapped element as the candidate as non-normative in **step 4** of [selecting the best candidate](https://drafts.csswg.org/css-nav-1/#select-the-best-candidate).

**Comment by hugoholgersson** [comment]
Do Chromium's layout tests pass with this change?

**Comment by jeonghee27** [comment]
> Do Chromium's layout tests pass with this change?
Yes, It did on https://chromium-review.googlesource.com/c/chromium/src/+/1905297

I'am planning to add a bug page and resolve conflict soon.

**Comment by jeonghee27** [comment]
I'm trying to apply new direction rule to chromium likes below. (https://chromium-review.googlesource.com/c/chromium/src/+/1905297/14/)

**AS-IS**
```
// if both edges of |a| are below the respective ones of |b|.
static inline bool Below(const PhysicalRect& a, const PhysicalRect& b) {
  return a.Y() >= b.Bottom() || (a.Y() >= b.Y() && a.Bottom() > b.Bottom() &&
                                 a.X() < b.Right() && a.Right() > b.X());
}
// For overlapping rects, |a| is considered to be on the right of |b|
// if both edges of |a| are on the right of the respective ones of |b|.
static inline bool RightOf(const PhysicalRect& a, const PhysicalRect& b) {
  return a.X() >= b.Right() || (a.X() >= b.X() && a.Right() > b.Right() &&
                                a.Y() < b.Bottom() && a.Bottom() > b.Y());
```
**TO-BE**
```// For overlapping rects, |a| is considered to be below |b|,
// if the top edge of |a| is below the top edge of |b|.
static inline bool Below(const PhysicalRect& a, const PhysicalRect& b) {
  return a.Y() >= b.Bottom() ||
         (a.Y() > b.Y() && a.X() < b.Right() && a.Right() > b.X());
}

// For overlapping rects, |a| is considered to be above |b|,
// if the bottom edge of |a| is above the bottom edge of |b|.
static inline bool Above(const PhysicalRect& a, const PhysicalRect& b) {
  return a.Bottom() <= b.Y() ||
         (a.Bottom() < b.Bottom() && a.Right() > b.X() && a.X() < b.Right());
}

// For overlapping rects, |a| is considered to be on the right of |b|,
// if the left edge of |a| is on the right of the left edge of |b|.
static inline bool RightOf(const PhysicalRect& a, const PhysicalRect& b) {
  return a.X() >= b.Right() ||
         (a.X() > b.X() && a.Y() < b.Bottom() && a.Bottom() > b.Y());
}

// Return true if rect |a| is on the left of |b|. False otherwise.
// For overlapping rects, |a| is considered to be on the left of |b|,
// if the right edge of |a| is on the left of the right edge of |b|.
static inline bool LeftOf(const PhysicalRect& a, const PhysicalRect& b) {
  return a.Right() <= b.X() ||
         (a.Right() < b.Right() && a.Bottom() > b.Y() && a.Y() < b.Bottom());
}```


But @hugoholgersson seems to have a different opinion.
I hope to make agreement here.

In below picture, 
![image](https://user-images.githubusercontent.com/34676178/72042688-8e5aa100-32f2-11ea-9de2-430065238037.png)

**Let's assue we press DOWN key on "A", "C", "E"**
1. "B" is next target from A?
2. "D" is next target from C?
3. "F" is next target from E?

**Let's assue we press UP key on "B", "D", "F"**
4. "A" is next target from B?
5. "C" is next target from D?
6. "E" is next target from F?

In my opinion and by spec, SpatNav consider only top edge when user press down.
So 1~3, 5~6 are wrong.


@hugoholgersson 
I wonder, your idea about the rules.

**Comment by hugoholgersson** [comment]
Thanks for the example. It shows that Above() and LeftOf() are needed to avoid loops such as [1]. I think you should emphasize this in a code comment and perhaps also in a spec comment.

[1] down, down, down: A, B, A

We avoid [1] thanks to Below()'s `a.Y > b.Y`. `>=` would cause [1].

(Please correct me if I'm wrong.)

---

By the way, why don't we relax the condition on the orthogonal axis? For example:

```
Below() {
  a.Y >= b.Bottom ||
    (a.Y > b.Y && a.X <= b.Right && a.Right >= b.X)
}
```

The current wording does not prevent `<=` and `>=` on the X-axis. [2] [3] 

[2] _"top edge is below the top edge of searchOrigin’s boundary box if dir is down"_
[3]  _"whose boundary box partially overlaps with inside area of searchOrigin"_

**Comment by jeonghee27** [comment]
> [1] down, down, down: A, B, A
We avoid [1] thanks to Below()'s a.Y > b.Y. >= would cause [1].

Yes right.


> why don't we relax the condition on the orthogonal axis? For example:
> [2] "top edge is below the top edge of searchOrigin’s boundary box if dir is down"

Do you think it is overlap even if 1px is attached?
I didn't. but It also makes sense. But it's a bit confusing to tell 1px overlap.
just confusing... I'll fix patch. 
I hope there are no different opinion elsewhere  in the formula.

**Comment by hugoholgersson** [comment]
> Do you think it is overlap even if 1px is attached?

What do you mean by "attached"?

**Comment by jeonghee27** [comment]
![image](https://user-images.githubusercontent.com/34676178/72301129-2b2d8d80-36a9-11ea-8e02-e0cce411a37a.png)

this case, only 1px border looks overlapped, (a.X = b.Right)
I meant this case.

**Comment by hugoholgersson** [comment]
Good pictures!

> Do you think it is overlap even if 1px is attached?

Yes. To me, [2] is overlapping but [1] is not.

> I thought we didn't need to include these two cases as overlap cases.

Yeah, I don't what's best either, but I'd say [2] should be considered "overlapping" even though only 1px overlaps. 1px gets bigger with zooming. :) 

>  I wonder your opinion

With a.X <**=** b.Right, we include [2] but not [1] so let's use **<=** in the formula? 

Otherwise, I think the formula looks good! 

When implementing this in Chrome, you will need to use `PhysicalRect::IntersectsInclusively`, not `PhysicalRect::Intersects`.

**Comment by jeonghee27** [comment]
Thanks for your guide. and I understand your opinion.
I'll use PhysicalRect::IntersectsInclusively and fix "a.X <= b.Right," in this week.
then, I'm going to ping you on chromium review page.
